packages/base/src/defineElement.ts

```diff
@@ -0,0 +1,9 @@
+export const defineElement = (elementName: string, elementConstructor: CustomElementConstructor, options?: { extends: string; }): void => {
+    // check for element existence
+    if (customElements.get(elementName)) return;
```

Comment:
Relying on something like this at the library level leaves a number of important questions and concerns found in #1024 unanswered. As you've used a pattern similar to this in the past, maybe you can help to answer them?

1. As a consumer who ends up with multiple versions of the same package in their project without my knowledge, I'd like to be made aware of this situation so that I can further optimize the code I deliver to my end users. With this gate, subsequent registration of the same element will be squelched and other than doing a deep dive of their built code or project lock, how do you see such a consumer discovering this situation?
2. As a consumer who chooses to use dependencies with mismatched package versions, I'd like to be able to know what version actually gets preferred by my build/application. As this technique prefers the first registration, how can a consumer know that this pattern has registered the one they'd actual like registered? This can be particularly useful when leveraging bundled code.
3. As a consumer who chooses to leverage multiple copies of the same version of a package in a projects, I'd like to be made aware so that I can reduce the code I ship to my users. How would a consumer in this case know that they've been building the code they deliver to their end users appropriately in this context? If I understand correct, this is very close to your use case and is an important goal to keep in mind. JS is the MOST expensive asset that we can ship to our users and patterns like this that can conceal how we're handing cost like this to our users can be a foot gun to longer term performance goals.
4. As a library vendor, I'd like to deliver quality and performance improvements to my consumers with the lowest likelihood of breaking changes. I may be overly optimistic that any later change to this can be made without it being breaking, but, particularly as this related the number 2 above, WHEN we move away from this pattern in the future I don't see a way to do it that isn't breaking. When something like Scoped Registries releases the code order prioritization of this approach there will be elements on a page that suddenly do not work the way they did before.
5. As a consumer of custom elements, I'd like to know when the libraries I leverage have diverged from the specification so I can rationalize what that means for my project. How can we document the decisions required to satisfy the above questions so that we can place consumers on the best track towards success? The best location I can think of just now is https://opensource.adobe.com/spectrum-web-components/getting-started. If we can come to an agreement on this, adding those findings to a section on that page would likely be enogh for now. Longer term does expanding that into a "Developing with SWC" section seems like a good idea?

Reply:
I understand wanting the library to offer opportunities for consumers to catch potential mistakes and opportunities for optimization, but having the library crash does not seem like the right way to do that. Could we guard against crashing and add a warning in the console? There are [cases] beyond version mismatches (https://wiki.corp.adobe.com/display/assetscollab/Express+Icons+Strategy) where we may want to reuse a common namespace